### ISSUE TITLE
destination_in directive

### ISSUE THREAD

**Issue by foxcpp** [foxcpp/maddy#138]
# Use case

Having mailboxes on two servers, one of them being maddy. Some mailboxes are handled by maddy, some are by upstream server. This configuration seems to be needed rather often, according to what I observe on #postfix IRC channel. Due to the maddys ability to do transparent forwarding (#33), it is not prone to backscatter problems and thus can be used safetly.

However, in this configuration, there should be a way to tell maddy which mailboxes should be handled by it and which should be handled by upstream. It is better to have these lists in a DB or at least in on-disk hash table (generated from separate file).

# Proposed solution

```
destination_in MODULE_REFERENCE {
  deliver_to smtp_upstream smtp://192.168.0.245
}
```

Due to implementation details, `destination_in` will always take precedence over `destination` directives.

Additionally, there is a problem with that particular syntax since it can't be used with inline definitions (e.g. `destination_in file_list { } { }` is not a valid config syntax).

Module needs to implement AddressList interface:
```go
type AddressList interface {
   CheckAddress(rcptTo string) bool
}
```
Can be implemented by sql module for DB-based lookups or by "file_list" for file-based lookups.
Also this interface can be used as a part of #56 implementation.

## Configuration example for use case

```
file_list remote_addrs {
  path /etc/maddy/remote_addrs.list
}

destination_in remote_addrs {
  deliver_to smtp_downstream smtp://192.168.0.234
}
destination example.org {
  deliver_to local_mailboxes
}
```

_Adding to 0.1 milestone since this can affect core design._

**Comment by foxcpp** [comment]
Dropped from 0.1 milestone. Given that we are staying with current pipeline design, it can be implemented post-0.1.

**Comment by foxcpp** [comment]
Additional use case: If authentication providers also implement the mentioned interface, this combined with replace_rcpt and auth_perdomain can be used to implement catch-all addresses.
Here is the example:

```
destination_in local_authdb {
  deliver_to local_mailboxes
}
destination myDomain.com {
  modify {
    replace_rcpt /.+/ default@myDomain.com
  }
  deliver_to local_mailboxes
}
```
